### ISSUE TITLE
Reporting Project - Complete Update to reporting system

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#47]
In GitLab by @kdjstudios on Oct 24, 2016, 24:43

**Submitted by:** Gary

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/2019-05-16-39675

**Server:** All

**Client/Site:** All 

**Account:** All

**Issue:** 
 
This ticket is being used for the parent ticket to review updating SAB to using a new/external reporting system. This new system we will want to be able to accomplish all of the requested below in. 

List of existing Gitlab tickets:
* #796
* #485

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 17, 2019, 08:46

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 17, 2019, 09:13

Gary wrote:

> Do we have the ability to schedule report delivery out of SAB?
> 
> Full message: [original_message__4_.html](/uploads/f9a5d5176ce7b2c443668df52b59858e/original_message__4_.html)

I wrote:

> Currently there is no functionality of setting up schedules for reports in SAB. We already have an existing ticket for this which is currently in part of the wishlist items and the reporting project.  


Gary Wrote:

> How long would it take to make this a reality.
> 
> Pre MVP - Ability to schedule the reports Cori describes below.
1. Weekly Aging Reports run after midnight on Sunday. (Emailed to NCSM Billing)
1. The Accounts Receivable as Invoiced reports run after the Site billing cycle is closed.
> 
> MVP - Engine to allow the scheduling and delivery of certain reports by the tech team.  In other words we get a request like this and we have a system where the tech team can go on, define the report and its parameters and set the schedule
*  This should include specifying a timezone for the schedule (behind the scenes we do all UTC).
*  Daily, Weekly, Monthly, hour of specified days
*  Reports will be emailed to the specified address.
> 
> MVP v2 - Engine to allow the scheduling and delivery of certain reports by the users.
* Users can see any report they have scheduled and they can schedule any report they have access to.
* Admins can see all report schedules and can schedule anything they like.

I wrote:

> Thank you, I have forwarded this to the team to provide an updated estimate and timeline.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 17, 2019, 09:14

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Aug 8, 2019, 05:02

@kdjstudios 

Can we step further in this ticket to provide an estimate?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 8, 2019, 08:52

@pchaudhary @tim.traylor 

I would believe so, as it has been marked as needing an estimate now for 2 months. But please discuss with @tim.traylor to confirm, I know while I was on vacation things changed priorities and I don't want to contradict anything.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 4, 2019, 14:12

Gary wrote:

> Kyle - I sent in a request for a time estimate on May 17 where I specifically defined the Pre MVP, MVP and MVP v2.
> 
> When was this "put on hold" and why wasn't I notified?  Who make this decision to not respond to me and to put my request on hold?
> 
> This is yet another example of why I am frustrated.  As the business owner of the product I am the one who is supposed to be setting priorities!!!!

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Sep 6, 2019, 05:00

@kdjstudios 

> Pre MVP - Ability to schedule the reports Cori describes below.
>
> 1. Weekly Aging Reports run after midnight on Sunday. (Emailed to NCSM Billing)
>
> 1. The Accounts Receivable as Invoiced reports run after the Site billing cycle is closed.

* We are not sure about the workflow but if it only needs a backend cron job which will run and send the results as an email, then it would take around 1 day (8 hrs).


* We still need the clarification on the Weekly Aging reports selected options according to which it will run.

* We also need to know the Timing after which the Account Receivable reports will run. Do we need to run it just after the Billing Cycle close or weekly, daily to check if the billing cycle is closed or not?

* Please note, the above estimate is to get a rough idea about the estimate. We will post a new estimate after the full clarification of the functionality.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 6, 2019, 08:21

@amishra 

Please bring this up for discussion with @tim.traylor and @jredfield in your morning meetings for the clarifications.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Sep 6, 2019, 08:31

@kdjstudios @jredfield Hi Kyle, Pls pass along the 8 hour estimate to Gary on setting up the two priority reports and let him know the additional estimates will follow.  He may want to approve the 8 so that can be completed.  Thx

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 6, 2019, 08:49

I wrote:

> We escalated this and the Aavaz support team was able to provide an initial estimate of 8 hours for the PRE-MVP topic to schedule the two priority reports. The two additional estimates will follow once the Aavaz team has provided them accordingly. Please let us know if this is approved for the team to begin working on the PRE-MVP.

Gary wrote:

> I need the entire estimate before making this call.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 11, 2019, 04:05

@kdjstudios 

Thanks for the update. 

We need the clarification on some points to provide you the estimate.

> Pre MVP - Ability to schedule the reports Cori describes below.
>
> 1. Weekly Aging Reports run after midnight on Sunday. (Emailed to NCSM Billing)
>
> 1. The Accounts Receivable as Invoiced reports run after the Site billing cycle is closed.

* The cronjob will run to schedule the reports on the backend & the result will mail to the requested email address, right?
* We also need the selected options for both of the reports according to which it will run. **eg- Form, Include Draft**
* We also need to know the Timing after which the cron for **Account Receivable Report** will run. Do we need to run it daily/weekly to check if the billing cycle is closed or not?


> MVP - Engine to allow the scheduling and delivery of certain reports by the tech team.  In other words, we get a request like this and we have a system where the tech team can go on, define the report and its parameters and set the schedule
>
> - This should include specifying a timezone for the schedule (behind the scenes we do all UTC).
>
> - Daily, Weekly, Monthly, the hour of specified days
>
> - Reports will be emailed to the specified address.

* We had a discussion with [Sumeet] ( @smasih ) & he told us that as per the discussion held on 10^th September 2019, we need to get a design first & according to that design we will provide an estimate.

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Sep 11, 2019, 05:55

Gary is requesting a design for this process which he wants to approve first before work commences.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 11, 2019, 09:42

@jredfield 

Lets setup a call and review @pchaudhary concerns above and also we can then setup any mock ups or design documents as needed. Please give me an IM or email when your available. Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 12, 2019, 09:46

This was mentioned on today's call the Aavaz team did have some suggestions and ideas on how to set this up. We asked them to provide them in the ticket and we will then use those to also review with Gary and his expectations.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 2, 2019, 13:01

## Mock Up:

### View Schedules Page:

This page will list off all the current schedules. Non-Admin users will only be able to see/edit their own schedules. while Admin users can see/edit all schedules for all users. 

**Buttons:**
* New Schedule: This will navigate to the report schedule page mocked up below and allow for a user to create a new schedule.
* Run Now: This will execute the schedule immediately. Can be used to test or send a report that was missed?
* Edit: This will navigate to the report schedule page and pre-load the settings specified on the selected schedule. 
* Pause/Resume: This will toggle if the schedule is active. Non-active schedules will not run, but will stay in the user interface.
* Delete: This will completely remove the schedule. We will need a pop up confirmation to allow the user to confirm. This will say "You are deleting this schedule. This action can not be undone. Please confirm."
* History Log: This will show the history log of when the report was ran and who it was delivered to at when as shown in the mock up below.

![View_Report_Schedules](/uploads/fbaa11677e5fb3c455efb9ff28021650/View_Report_Schedules.png)


### Report Schedule:

This page will allow for a user to schedule any report and specify who it is delivered to.

**Sections:**

* Report drop-down and options: At the top of the page we will allow for a user to select which report. once that is selected, the reports options and parameters will be loaded below for the user to specify accordingly.
* Schedule: This section will allow for scheduling the report via daily, weekly, monthly, or yearly. The screenshots of each option and its sub options I took from the outlook scheduling tool. Each option will display its appropriate sub options when selected. 
* Start and End date: This will allow for the user to specify when the schedule is to start and when it is to stop. When a schedule stops, it simply becomes non-active. 
* Cancel: This will cancel any updates and return to the view schedules page.
* Save: This will save any updates or changes to the schedule and return to the view schedules page. 

![UI_for_Report_Schedules](/uploads/ff00c19f7b1d8f92fb845b2eb325559f/UI_for_Report_Schedules.png)

### Report History Log

This page will show the tracking of the report schedule and all execution or modification to it. 

![Report_Schedule_History_Log](/uploads/2db3b8f41c8dd5fb5c6a17bf7bacd6e0/Report_Schedule_History_Log.png)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 2, 2019, 13:07

@amishra @smasih 

May we please get an estimate for the above scheduling tool for the reporting. To our understanding the MVP and MVP2 are the same core functionality, it is just allowing user access to this, verse only admin access. So there should be little difference in those.  We already have the 8 hour estimate to simply manually schedule two reports. so we will need the MVP and MVP2 estimates to provide to Gary.

Also if you have any questions or concerns about the functionality mocked up, please let us know.

@jredfield 

May you please provide an estimate of what you think this should take also.

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Oct 17, 2019, 12:56

@amishra @smasih @kdjstudios I'm estimating 12 hours for the MVP (tech team), and 4 hours for the MPV2 task (users).  The MVP2 should be much quicker due to its similarity to MVP 1. So 16 hours total for MVP and MVP2.

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Oct 21, 2019, 17:43

On the top mockup (View Schedules Page: http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/47#note_48627) it might make more sense to have the five buttons in one place so we don't have to keep duplicating them in each row.  Perhaps they could be placed right above the header row of the grid.

| New Schedule | Run Now | Pause/Resume | Edit | History Log | Delete |

[ Column headers here ... ]

Then the user would select the row with the desired report and click one of the buttons.  This would also allow more rows per page since they wouldn't have to be so high to accomodate all the buttons.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 22, 2019, 09:38

@jredfield 

I do like the idea of having those buttons only once, Would it effect anything while scrolling the page? Or would they 'float' at the top of the page no matter where the user scrolls?

Also, we could simply just keep the buttons in a single row rather then a grid like they currently are to also save space if that is needed. We will want to have Gary's input on what he wants.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 29, 2019, 09:50

This was reviewed today with Gary and he would like for us to provide an estimate for prioritization and approval. @jredfield has provided his initial estimate for 16 hours in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/47#note_49469.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 11, 2019, 03:34

@kdjstudios @jredfield 

Thank you for the Update.

## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|X1-Large|30|Worked with Concepts, implementation ok|38|Low Impact|7.0|45|

**Bifurcation of estimate:**

1. Cron Engine  - `5 hours`
1. Front-end + Back-end + validations + report  - `33 hours`
1. Testing & Debugging  - `7 hours`

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 11, 2019, 11:05

@amishra @smasih 

Just a note, as I keep seeing this term used in tickets incorrectly. You may need to check the term "Bifurcation", as I believe you are using it incorrectly. I believe the term you are needing to use is "Explanation".

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Nov 19, 2019, 13:48

@kdjstudios @amishra @smasih 

Kyle wrote
> Just a note, as I keep seeing this term used in tickets incorrectly. You may need to check the term "Bifurcation", as I believe you are using it incorrectly.

"Bifurcation" means the division of something (e.g. a task in a ticket) into two parts.  That stems from "Bi", meaning two (e.g. BIcycle, BIfocal, BInary). 

For the above task estimate, I don't see anything based on "two".  It is one task, and the breakdown of the estimate has *three* parts.  (Just wearing my "Grammar Police" cap).

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Nov 19, 2019, 14:07

@kdjstudios @amishra We still want to review this some more with Gary, and he may want to re-prioritize it.

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Dec 4, 2019, 13:13

@kdjstudios @amishra @smasih This is being put on hold until we explore other reporting options, Gary's concern is that we are going to add another week's worth of programming effort to add a feature on top of a very ad-hoc, increasingly unmanageable reporting "system". Some of us met with Guy the "Data Guy" who is going to expose SAB/mongoid data in a way that we can have it serve as a centralized data source for a reporting engine, complete with scheduler and other features we want.  We have narrowed it down to two choices: Jasper (Jaspersoft) and Pentaho 8. There are pros and cons with both of these systems. Gary expects an answer by beginning of next week (12/9/19). If we choose a reporting engine at that time, this current estimate will not be relevant anymore, and we will probably introduce a new ticket for the overhauled reporting system. @tim.traylor will also be able to contribute significantly to this overhaul.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 4, 2019, 14:32

Thanks @jredfield 

Marking this into pause until we hear back on which reporting system/engine is chosen.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 10, 2019, 09:05

This was briefly mentioned today and will be further discussed during the managers meeting later..

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 25, 2020, 13:32

@jredfield May you follow up with Tim and Kess to setup a meeting so we can start making progress on this. Hopefully we can at minimum decide on which product/system we will be using and then start to work out a project plan from there.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 25, 2020, 13:32

assigned to @jredfield

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 25, 2020, 15:41

changed title from **Re{-quest Ability to Schedule Reports-}** to **Re{+porting Project - Complete Update to reporting system+}**

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 25, 2020, 15:41

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 25, 2020, 15:42

changed the description
